### PR TITLE
PARQUET-418: Refactored parquet_reader utility for printing file contents.

### PR DESCRIPTION
**Pull request by asandryh**
This pull request contains the following changes:
* Modified parquet_reader utility: refactored, fixed memory leaks, merged compute_stats utility to get rid of code duplication.
* Added a flag --only-stats to parquet_reader to print only file statistics.
* Modified InMemoryInputStream to own its buffer.

All the code repetition still remaining in parquet_reader clearly highlights the need for specialized ColumnReader classes. I will create a new JIRA for this improvement.

### CODE REVIEW

File: src/parquet.cc [hunk 1]
```diff
@@ -33,8 +33,20 @@ using parquet::PageType;
...
 
-InMemoryInputStream::InMemoryInputStream(const uint8_t* buffer, int64_t len) :
-  buffer_(buffer), len_(len), offset_(0) {
+InMemoryInputStream::InMemoryInputStream(int64_t len) : len_(len), offset_(0) {
+  buffer_ = (uint8_t*)malloc(sizeof(uint8_t) * len_);
+}
+
+InMemoryInputStream::~InMemoryInputStream() {
+  free(buffer_);
+}
```
Comment (wesm) [inline src/parquet.cc]: I think it would be preferable to leave the management of this memory to parquet_reader.cc; the intent of this class was to implement InputStream on any contiguous data in virtual address space, so by removing the `buffer` parameter it is no longer able to work with memory regions that are not allocated and freed by the class instance.

File: example/parquet_reader.cc [hunk 2]
```diff
@@ -41,254 +39,227 @@ static string ByteArrayToString(const ByteArray& a) {
...
+        if (col.meta_data.__isset.dictionary_page_offset &&
+            col_start > col.meta_data.dictionary_page_offset) {
+          col_start = col.meta_data.dictionary_page_offset;
         }
-        case Type::BYTE_ARRAY: {
-          ((ByteArray**)column_ptr)[c] =
-              (ByteArray*)malloc(sizeof(ByteArray) * INIT_SIZE);
-          type_array[c] = Type::BYTE_ARRAY;
-          break;
+        std::auto_ptr<InMemoryInputStream> input(
```
Comment (wesm) [inline example/parquet_reader.cc]: Note `std::auto_ptr` is deprecated in C++11 in favor of `std::unique_ptr`

File: src/parquet.cc (hunk 1 above)
Comment (asandryh) [inline src/parquet.cc]: IMHO, it's dangerous to manage the life cycle of a memory buffer outside of the object that's using it. Anyway, instead of modifying InMemoryInputStream, I can add another InputStream that manages its own memory.

File: example/parquet_reader.cc (hunk 2 above)
Comment (asandryh) [inline example/parquet_reader.cc]: Right. When we switch to C++11, we'll replace it with `unique_ptr`.

File: example/parquet_reader.cc [hunk 3]
```diff
@@ -41,254 +39,227 @@ static string ByteArrayToString(const ByteArray& a) {
...
+    } else {
+      filename = argv[i];
     }
-
-    return column_ptr;
   }
 
-  fclose(file);
-  return NULL;
+  readParquet(filename, printContents);
```
Comment (wesm) [inline example/parquet_reader.cc]: It would be great if we could turn this into a `ParquetFileReader::DebugPrint` on top of #20, then the parquet_reader.cc executable would only be a few lines of code.

**Comment by wesm** [comment]
@nongli this patch is non-overlapping with #19 and #20. I suggest we merge all 3 patches once you review and I'll make some follow up patches to continue library-ifying things.

File: example/parquet_reader.cc (hunk 3 above)
Comment (asandryh) [inline example/parquet_reader.cc]: It is reasonable to convert parquet_reader.cc into a ParquetFileReader method. For simplicity, let's do this in a separate PR, after this #18 and #20 are completed.

**Comment by wesm** [comment]
lgtm, thank you

**Comment by asandryh** [comment]
@nongli Can you pls take a look at this patch and provide feedback (or pull this into parquet-cpp if you're ok with everything)?

**Comment by wesm** [comment]
@asandryh This patch needs to be rebased on upstream master before it can be merged. There are multiple merge commits here so unfortunately rebasing is a bit tricky

**Comment by wesm** [comment]
In the future you will want to avoid merging master into a branch; in the case of conflict you can rebase.

**Comment by wesm** [comment]
I prepared a rebased commit here

https://github.com/wesm/parquet-cpp/commit/f65424bbc9685c2682f7adb69e0da691247546ad

If this happens in the future, you can use `git merge --squash` to stage the changes in the git index and then create a new commit with only the diff from the original branch. 

```
git merge --squash asandryh/PARQUET-418 
git commit --author="Aliaksei Sandryhaila <aliaksei.sandryhaila@hp.com>" -m "[PARQUET-418]: Added/modified a utility for printing a file contents."
```

**Comment by wesm** [comment]
Please see #23. @nongli I refactored parquet_reader.cc into an instance method of `ParquetFileReader`. Since #23 includes this commit, I suggest for expediency that you review first #22 (PARQUET-433) and then #23 (PARQUET-451 and PARQUET-453)

File: src/parquet/parquet.h [hunk 4]
```diff
@@ -117,6 +117,22 @@ class InMemoryInputStream : public InputStream {
...
+  ScopedInMemoryInputStream(const int64_t len);
+  ~ScopedInMemoryInputStream();
+  uint8_t* data();
+  int64_t size();
+  virtual const uint8_t* Peek(int num_to_peek, int* num_bytes);
+  virtual const uint8_t* Read(int num_to_read, int* num_bytes);
+
+ private:
+  int64_t len_;
+  uint8_t* buffer_;
```
Comment (nongli) [inline src/parquet/parquet.h]: use std::vector<uint8_t>
and
scoped_ptr<InMemoryInputStream>.

File: src/parquet/parquet.h [hunk 5]
```diff
@@ -117,6 +117,22 @@ class InMemoryInputStream : public InputStream {
   int64_t offset_;
 };
 
+// A wrapper for InMemoryInputStream to manage the memory.
+class ScopedInMemoryInputStream : public InputStream {
+ public:
+  ScopedInMemoryInputStream(const int64_t len);
```
Comment (nongli) [inline src/parquet/parquet.h]: Comment 'len'

File: example/parquet_reader.cc [hunk 6]
```diff
@@ -41,254 +39,229 @@ static string ByteArrayToString(const ByteArray& a) {
...
 }
 
+string type2String(Type::type t) {
+  switch(t) {
+    case Type::BOOLEAN:
+      return "BOOLEAN";
+      break;
+    case Type::INT32:
+      return "INT32";
+      break;
```
Comment (nongli) [inline example/parquet_reader.cc]: you have some redundant breaks

**Comment by nongli** [comment]
Can you address the comments in parquet.h/.cc and we can deal with anything else as a follow up?

Also, can you update the title of the PR to be "PARQUET-XXX: ..." (remove "Fixed")

File: example/parquet_reader.cc (hunk 6 above)
Comment (asandryh) [inline example/parquet_reader.cc]: You are right, semantically these breaks are redundant. However, some compilers may complain about missing breaks in a switch statement, so it's safer to keep them.

File: src/parquet/parquet.h (hunk 5 above)
Comment (asandryh) [inline src/parquet/parquet.h]: This is the length of the buffer allocated by ScopedInMemoryInputStream.
@wesm Since you have incorporate this PR into #23, can you please add a comment to #23?

File: src/parquet/parquet.h (hunk 4 above)
Comment (asandryh) [inline src/parquet/parquet.h]: It is redundant. buffer_ is deallocated in the destructor.

**Comment by asandryh** [comment]
@nongli Thank you for the comments!

File: src/parquet/parquet.h (hunk 4 above)
Comment (wesm) [inline src/parquet/parquet.h]: I think he is saying to use `std::vector` for memory allocation/deallocation rather than `malloc / free`

Comment (asandryh) [inline src/parquet/parquet.h]: malloc is perfectly fine in this case. std::vector will call it, internally, too; there is no need to obscure this memory allocation. And we don't need vector or a smart pointer to manage memory deallocation - this is done in ~ScopedInMemoryInputStream() dctor.

Comment (wesm) [inline src/parquet/parquet.h]: Well, the point is, why re-implement RAII in this class when we already have an STL class available that does this for us?

Comment (nongli) [inline src/parquet/parquet.h]: I think as a matter of coding style, we should eliminate as many uses of malloc/news as possible. 

We can just remove the dtor, the owernship is self documenting. I believe the code right now is correct but if the construction of the stream in the ctor throws, buffer_ will be leaked. Using RAII, this problem would not exist.

Comment (wesm) [inline src/parquet/parquet.h]: +1 -- I'm still acclimatizing to modern C++ (and thus have made a couple `new` mistakes, e.g. https://github.com/wesm/parquet-cpp/commit/0714c473f5ea7eaa0f6c5c4055028948951f93b9) so will endeavor to follow these guidelines.

Comment (asandryh) [inline src/parquet/parquet.h]: That's fine, let's use std::vector<char>.

File: src/parquet/parquet.h (hunk 5 above)
Comment (wesm) [inline src/parquet/parquet.h]: Sure, I will do that

File: src/parquet/parquet.h (hunk 4 above)
Comment (wesm) [inline src/parquet/parquet.h]: If you can make this change, I will rebase my patches on and get a passing build so we can move along the patch queue

Comment (wesm) [inline src/parquet/parquet.h]: @nongli I can fix this in my downstream patches since I already have a rebased version of this patch in #23

Comment (asandryh) [inline src/parquet/parquet.h]: Will do shortly.

**Comment by wesm** [comment]
@asandryh this branch has gotten pretty messed up. here are some commands to help fix it

```
git remote add wesm git@github.com:wesm/parquet-cpp.git
git fetch upstream  # presumes this is apache/parquet-cpp.git
git fetch wesm
git reset --hard upstream/master
git cherry-pick 7279b20   # this is the rebased commit from #23
git push -f
```

**Comment by wesm** [comment]
If you ever screw up, find your last commit in `git reflog` and run `git reset --hard HASH`

**Comment by asandryh** [comment]
@wesm Thanks! I tried to avoid hard reset by squashing my commits, but it didn't work.

**Comment by wesm** [comment]
@asandryh Travis CI builds on the apache org are super slow this week. If you can enable Travis CI for asandryh/parquet-cpp and post verified builds from your branch that will help @nongli verify the commit to merge faster

**Comment by asandryh** [comment]
I've enabled Travis and confirm that this PR builds successfully.

**Comment by wesm** [comment]
OK -- can you make the std::vector change then post a link to the verified build so @nongli can merge? Then I will rebase my patches

File: CMakeLists.txt [hunk 7]
```diff
@@ -133,7 +133,7 @@ message(STATUS "Boost libraries: " ${Boost_LIBRARIES})
 
 # find thrift headers and libs
 find_package(Thrift REQUIRED)
-include_directories(SYSTEM ${THRIFT_INCLUDE_DIR})
+include_directories(SYSTEM ${THRIFT_INCLUDE_DIR} ${THRIFT_INCLUDE_DIR}/thrift)
```
Comment (nongli) [inline CMakeLists.txt]: Why did you change this? Is this to change #include <thrift/xyz> to #include<xyz>

We should namespace include paths as much as possible.

Comment (asandryh) [inline CMakeLists.txt]: This ensures that parquet-cpp compiles with older versions of Thrift.

**Comment by asandryh** [comment]
Successful Travis build: https://travis-ci.org/asandryh/parquet-cpp/builds/105048491

File: example/example_util.h [hunk 8]
```diff
@@ -17,7 +17,28 @@
...
+  }
+  ~InputFile() {
+    if (file != NULL) {
+      fclose(file);
+    }
+  }
+
+  FILE* getFileHandle() { return file; }
+  bool isOpen() { return file != NULL; }
+  std::string getFilename()  { return filename; }
```
Comment (nongli) [inline example/example_util.h]: this should be const std::string&

File: src/parquet.cc [hunk 9]
```diff
@@ -48,7 +47,32 @@ const uint8_t* InMemoryInputStream::Read(int num_to_read, int* num_bytes) {
...
+
+const uint8_t* ScopedInMemoryInputStream::Read(int num_to_read,
+                                               int* num_bytes) {
+  return stream_->Read(num_to_read, num_bytes);
+}
+
+
 ColumnReader::~ColumnReader() {
+  delete stream_;
```
Comment (nongli) [inline src/parquet.cc]: This has unclear ownership semantics. We should make this a unique ptr too. Take a look at this example
http://en.cppreference.com/w/cpp/memory/unique_ptr

**Comment by nongli** [comment]
I'm going to merge this. Please address in follow ups. Thanks!

File: example/example_util.h (hunk 8 above)
Comment (wesm) [inline example/example_util.h]: this is removed in subsequent commits

File: src/parquet.cc (hunk 9 above)
Comment (wesm) [inline src/parquet.cc]: thanks, I will fix
